gin/hmac.go

```diff
@@ -0,0 +1,102 @@
+package gin
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+)
+
+var ErrInvalidSignature = errors.New("invalid signature")
+
+type StrFromCtx func(c *gin.Context) (string, error)
+
+// HmacDefaultSignatureHeader defines the default header name where clients should place the signature.
+const HmacDefaultSignatureHeader = "X-REQ-SIG"
+
+type HmacSha256Verifier struct {
+	keys       [][]byte
+	sigFN      StrFromCtx
+	sigEncoder func([]byte) string
+}
+
+func NewHmacSha256Verifier(keys []string) *HmacSha256Verifier {
+	keysB := make([][]byte, len(keys))
+	for i := range keys {
+		keysB[i] = []byte(keys[i])
+	}
+
+	return &HmacSha256Verifier{
+		keys: keysB,
+		sigFN: func(c *gin.Context) (string, error) {
+			return c.GetHeader(HmacDefaultSignatureHeader), nil
+		},
+		sigEncoder: func(b []byte) string {
+			return base64.StdEncoding.EncodeToString(b)
+		},
+	}
+}
+
+// WithSigFunction can be used to override signature location.
+// As a query string param for example:
+//
+//	func(c *gin.Context) (string, error) {
+//		return c.Query("sig"), nil
+//	}
+func (v *HmacSha256Verifier) WithSigFunction(sigFN StrFromCtx) *HmacSha256Verifier {
```

Comment:
hey, this kind of pattern may expose a partially built struct to the user

```
verifier := NewVerifier()

// use it in some way....

verifier.WithSignFn(...)

/// use it with new sig fn

verifier.WithSigEncoder(...)

// use it with new sig encoder
```

is this intended? If this usage is completely intended, then all good. Otherwise, prefer to change it to a builder pattern or functional options pattern